apstra/api_custom_collector.go

```diff
@@ -0,0 +1,257 @@
+	if err != nil {
+		return err
+	}
+
+	err = o.OsType.FromString(raw.OsType)
+	if err != nil {
+		return err
+	}
+
+	err = o.OsVersion.FromString(raw.OsVersion)
+	if err != nil {
+		return err
+	}
+
+	o.Model = raw.Model
+
+	for _, v := range strings.Split(raw.OsFamily, ",") {
+		var variant CollectorOSVariant
+		err = variant.FromString(v)
+		if err != nil {
+			return err
+		}
+		o.OsFamily = append(o.OsFamily, variant)
+	}
+
+	return nil
+}
+
+func (o *CollectorPlatform) MarshalJSON() ([]byte, error) {
+	var raw struct {
+		OsType    string `json:"os_type"`
+		OsVersion string `json:"os_version"`
+		OsFamily  string `json:"family"`
+		Model     string `json:"model"`
+	}
+	raw.OsType = o.OsType.String()
+	raw.OsVersion = o.OsVersion.String()
+	raw.Model = o.Model
+	raw.OsFamily = o.OsFamily[0].String()
+	for _, v := range o.OsFamily[1:] {
+		raw.OsFamily = raw.OsFamily + "," + v.String()
+	}
+	return json.Marshal(raw)
+}
+
+type Query struct {
+	Accessors map[string]string `json:"accessors"`
+	Keys      map[string]string `json:"keys"`
+	Value     string            `json:"value"`
+	Filter    string            `json:"filter"`
+}
+type Collector struct {
+	ServiceName             string
+	Platform                CollectorPlatform `json:"platform"`
+	SourceType              string            `json:"source_type"`
+	Cli                     string            `json:"cli"`
+	Query                   Query             `json:"query"`
+	RelaxedSchemaValidation bool              `json:"relaxed_schema_validation"`
+}
+
+// GetAllCollectors gets all the Collectors for all services
+func (o *Client) GetAllCollectors(ctx context.Context) ([]Collector, error) {
+	var response struct {
+		Items map[string]interface{} `json:"items"`
+	}
+	var collectors []Collector
+	// First Reach out to /collectors , we are interested really only in the keys, so that we can pull the collectors
+	err := o.talkToApstra(ctx, &talkToApstraIn{
+		method:      http.MethodGet,
+		urlStr:      fmt.Sprintf(apiUrlCollectors),
+		apiResponse: &response,
+	})
+	if err != nil {
+		return nil, convertTtaeToAceWherePossible(err)
+	}
+
+	for k := range response.Items {
+		cs, err := o.GetCollectorsByServiceName(ctx, k)
+		if err != nil {
+			return nil, err
+		}
+		for _, v := range cs {
+			v.ServiceName = k
+			collectors = append(collectors, v)
+		}
+	}
+	return collectors, nil
+}
+
+// GetCollectorsByServiceName gets all the Collectors that correspond to a particular service
+func (o *Client) GetCollectorsByServiceName(ctx context.Context, name string) ([]Collector, error) {
+	var ace ClientErr
+	var Response struct {
+		Items []Collector `json:"items"`
+	}
+	err := o.talkToApstra(ctx, &talkToApstraIn{
+		method:      http.MethodGet,
+		urlStr:      fmt.Sprintf(apiUrlCollectorsByServiceName, name),
+		apiResponse: &Response,
+	})
+
+	if err != nil {
+		err = convertTtaeToAceWherePossible(err)
+		if errors.As(err, &ace) && ace.Type() == ErrNotfound {
+			return Response.Items, nil
```

Comment:
I think this return is the "there are no collectors for that service name" case. Not an error.

Do we expect to have anything in `Response.Items` here? 

Feels like returning `nil, nil` would be more clear.

Reply:
If it's not found, there's no error, changed it to nil. nil
If there's a notfound error, it will be something else.